src/contracts/test/SmartOrder.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+pragma solidity ^0.7.6;
+pragma abicoder v2;
+
+import "../interfaces/GPv2EIP1271.sol";
+import "../interfaces/IERC20.sol";
+import "../libraries/GPv2Order.sol";
+import "../libraries/GPv2SafeERC20.sol";
+import "../libraries/SafeMath.sol";
+import "../GPv2Settlement.sol";
+
+contract SmartOrder is EIP1271Verifier {
+    using GPv2Order for GPv2Order.Data;
+    using GPv2SafeERC20 for IERC20;
+    using SafeMath for uint256;
+
+    bytes32 public constant APPDATA = keccak256("SmartOrder");
+
+    address public immutable owner;
+    bytes32 public immutable domainSeparator;
+    IERC20 public immutable sellToken;
+    IERC20 public immutable buyToken;
+    uint256 public immutable totalSellAmount;
+    uint256 public immutable totalFeeAmount;
+    uint32 public immutable validTo;
+
+    bool public funded;
+
+    constructor(
+        GPv2Settlement settlement,
+        IERC20 sellToken_,
+        IERC20 buyToken_,
+        uint32 validTo_,
+        uint256 totalSellAmount_,
+        uint256 totalFeeAmount_
+    ) {
+        owner = msg.sender;
+        domainSeparator = settlement.domainSeparator();
+        sellToken = sellToken_;
+        buyToken = buyToken_;
+        validTo = validTo_;
+        totalSellAmount = totalSellAmount_;
+        totalFeeAmount = totalFeeAmount_;
+
+        sellToken_.approve(
+            address(settlement.allowanceManager()),
+            totalSellAmount_.add(totalFeeAmount_)
+        );
+    }
+
+    modifier onlyOwner {
+        require(msg.sender == owner, "not owner");
+        _;
+    }
+
+    function fund() external onlyOwner {
+        require(!funded, "already funded");
+        funded = true;
+        sellToken.safeTransferFrom(
+            owner,
+            address(this),
+            totalSellAmount.add(totalFeeAmount)
+        );
+    }
+
+    function close() external onlyOwner {
+        uint256 balance = sellToken.balanceOf(address(this));
+        if (balance != 0) {
+            sellToken.safeTransfer(owner, balance);
+        }
+        selfdestruct(payable(owner));
```

Comment:
What happens if the owner is a contract and it doesn't have a function to receive eth?

Reply:
In any case, if this were a contract I deploy I'd prefer to have a `withdraw(token, amount)` function always callable at any time. :grin:

Reply:
> What happens if the owner is a contract and it doesn't have a function to receive eth?

It turns out, `selfdestruct` does not call the contract at all. As such, you can use this to get an Ether balance into a contract without a payable `fallback` (or `receive` in modern Solidity) function. Source: I tried the following in Remix:

```solidity
// SPDX-License-Identifier: CC0-1.0
pragma solidity ^0.8.0;

contract Balance {
    function val() external view returns (uint256) {
        return address(this).balance;
    }
}

contract Boom is Balance {
    receive() external payable {}
    function bang(address payable receiver) external {
        selfdestruct(receiver);
    }
}
```

And was able to use `Boom.bang()` to send Eth to a `Balance` contract that doesn't have a payabale `receive` function.

Reply:
Added `withdraw`.